### PR TITLE
Fix toolchain detection for worktree-local paths

### LINKED ISSUE #20177: Toolchains installed via nix are not detected for Python
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

See the [reproduction repository](https://github.com/WeetHet/zed-toolchains-bug-reproduction.git) for reproduction

In short, zed does not detect the accessible python toolchain when started from the Dock if they are installed via `nix` and accessible through `direnv`. It does detect them when started from the terminal

### Environment

Zed: v0.161.0 (Zed Nightly c41a8e33a0aa05708b6e21731e7e708a8f34e6c3)
OS: macOS 15.0.1
Memory: 8 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

Screenshot 2024-11-04 at 18 28 39

### PR DESCRIPTION
**Pull request by WeetHet**
Reimplements `pet::EnvironmentApi`, trying to access the `project_env` first
Closes #20177 

Release Notes:

- Fixed python toolchain detection when worktree local path is set

### CODE REVIEW

File: crates/project/src/toolchain_store.rs [hunk 1]
```diff
@@ -345,9 +350,11 @@ impl RemoteToolchainStore {
             Some(())
         })
     }
+
     pub(crate) fn list_toolchains(
         &self,
         worktree_id: WorktreeId,
+        _project_env: Option<HashMap<String, String>>, // TODO: use project_env on remote somehow?
```
Comment (osiewicz) [inline crates/project/src/toolchain_store.rs:357]: We should probably update `zed.proto` to pass project env, kinda like we do for tasks in https://github.com/zed-industries/zed/blob/3856599853b2a8aab15728847c737d13b08c3bfe/crates/proto/proto/zed.proto#L2187

Comment (WeetHet) [reply crates/project/src/toolchain_store.rs:357]: It's just weird, like, the environment is already from the remote, no? Why would I roundtrip it to the host and back?

File: crates/toolchain_selector/src/active_toolchain.rs [hunk 2]
```diff
@@ -113,9 +113,12 @@ impl ActiveToolchain {
                     .ok()?;
                 let toolchains = cx
                     .update(|cx| {
-                        project
-                            .read(cx)
-                            .available_toolchains(worktree_id, language_name, cx)
+                        project.read(cx).available_toolchains(
+                            worktree_id,
+                            todo!("worktree path"), // TODO: I don't know where to get this from?
```
Comment (osiewicz) [inline crates/toolchain_selector/src/active_toolchain.rs:118]: Prolly like `project.read(cx).worktree_for_id(worktree_id, cx)?.read(cx).abs_path()`

File: crates/project/src/toolchain_store.rs (hunk 1 above)
Comment (WeetHet) [reply crates/project/src/toolchain_store.rs:357]: Anyways, let's do this for now and leave this to deal with for remoting people for later

Comment (WeetHet) [reply crates/project/src/toolchain_store.rs]: Scratch that, I've come up with a better solution

File: crates/languages/src/python.rs [hunk 3]
```diff
@@ -427,6 +434,131 @@ impl ToolchainLister for PythonToolchainProvider {
     }
 }
 
+pub struct EnvironmentApi<'a> {
+    global_search_locations: Arc<Mutex<Vec<PathBuf>>>,
+    project_env: &'a HashMap<String, String>,
+}
+
+impl<'a> EnvironmentApi<'a> {
```
Comment (osiewicz) [inline crates/languages/src/python.rs:442]: This looks a bit like a copypaste of impl from pet itself; could we instead make `EnvironmentApi` a bit like a facade where it looks paths up in `project_env`, and if that fails, it uses `pet_core::os_environment::EnvironmentApi` (that can be it's another member)?

File: crates/project/src/toolchain_store.rs (hunk 1 above)
Comment (osiewicz) [reply crates/project/src/toolchain_store.rs]: It is better indeed :+1:

File: crates/languages/src/python.rs (hunk 3 above)
Comment (WeetHet) [reply crates/languages/src/python.rs:443]: Done

File: crates/languages/src/python.rs [hunk 4]
```diff
@@ -427,6 +434,100 @@ impl ToolchainLister for PythonToolchainProvider {
...
+        self.project_env
+            .get("HOME")
+            .or_else(|| self.project_env.get("USERPROFILE").cloned())
+            .map(|home| pet_fs::path::norm_case(PathBuf::from(home)))
+            .or_else(self.pet_env.get_user_home())
+    }
+}
+
+#[cfg(windows)]
+impl<'a> pet_core::os_environment::Environment for EnvironmentApi<'a> {
```
Comment (osiewicz) [inline crates/languages/src/python.rs:466]: These two cfgs differ just by `get_know_global_search_locations`, so ig we could #[cfg] at that functions level (have two different impls for that function and unify the rest).

Comment (WeetHet) [reply crates/languages/src/python.rs]: Yeah, I've just unified the implementations

**Comment by osiewicz** [comment]
Thanks!

**Pull request merged**